pipelined/bevy_pbr2/src/render/pbr.wgsl

```diff
@@ -346,17 +348,36 @@ fn omni_light(
     return ((diffuse + specular_light) * light.color.rgb) * (rangeAttenuation * NoL);
 }
 
-fn fetch_shadow(light_id: i32, homogeneous_coords: vec4<f32>) -> f32 {
-    if (homogeneous_coords.w <= 0.0) {
-        return 1.0;
-    }
-    // compensate for the Y-flip difference between the NDC and texture coordinates
-    let flip_correction = vec2<f32>(0.5, -0.5);
-    let proj_correction = 1.0 / homogeneous_coords.w;
-    // compute texture coordinates for shadow lookup
-    let light_local = homogeneous_coords.xy * flip_correction * proj_correction + vec2<f32>(0.5, 0.5);
+fn fetch_shadow(light_id: i32, frag_position: vec4<f32>) -> f32 {
+    let light = lights.omni_lights[light_id];
+
+    // because the shadow maps align with the axes and the frustum planes are at 45 degrees
+    // we can get the worldspace depth by taking the largest absolute axis
+    let frag_ls = light.position.xyz - frag_position.xyz;
+    let abs_position_ls = abs(frag_ls);
+    let major_axis_magnitude = max(abs_position_ls.x, max(abs_position_ls.y, abs_position_ls.z));
+
+    // do a full projection
+    // vec4 clip = light.projection * vec4(0.0, 0.0, -major_axis_magnitude, 1.0);
+    // float depth = (clip.z / clip.w);
+
+    // alternatively do only the necessary multiplications using near/far
+    let proj_r = light.far / (light.near - light.far);
+    let z = -major_axis_magnitude * proj_r + light.near * proj_r;
+    let w = major_axis_magnitude;
+    let depth = z / w;
```

Comment:
This is something that should be discussed. This bit trades a full matrix multiplication for these couple of lines, but at the cost of needing to update the shader if the light projection changes.

Because it's in the fragment shader it might be worthwhile as an optimization, but early optimization etc etc...

Reply:
I think this is reasonable for now, given that the projection needs to be a 90 degree field of view to fully cover the cube map. Any more would result on overlap, any less would result in gaps right?

Reply:
Yes. I'm also not sure what other projections make sense, except maybe infinite reverse.